File: reuse_go/athlete_interface/AthleteInterface.go
package athlete_interface

import "fmt"

type AthleteInterface interface {
	MySpeciality() string
}

type BaseAthlete struct {
	Speciality string
}

func (baseAthlete BaseAthlete) MySpeciality() string {
	return "My speciality is " + baseAthlete.Speciality + "."
}

type DistanceRunner struct {
	Firstname string
	Lastname  string
	Athlete   AthleteInterface
}

func (d *DistanceRunner) Display() {
	fmt.Printf("My name is %s %s. %v\n",
		d.Firstname,
		d.Lastname,
		d.Athlete.MySpeciality(),
	)
}

type Jumper struct {
	Firstname string
	Lastname  string
	Athlete   AthleteInterface
}

func (j *Jumper) Display() {
	fmt.Printf("My name is %s %s. %v\n",
		j.Firstname,
		j.Lastname,
		j.Athlete.MySpeciality(),
	)
}

File: reuse_python/src/athlete_interface.py
"""
Classes Reusability Project

This project demonstrates different ways to reuse classes in Python:
- Inheritance
- Composition
- Composition through interface implementation

Classes:
    Athlete: Represents an athlete with a specific specialty.
    Sprinter: Represents a sprinter, inheriting from Athlete.
    Thrower: Represents a thrower, utilizing Athlete class through composition.
    AthleteInterface: Interface defining the method for athlete specialty.
    AthleteBase: Class implementing the AthleteInterface.
    DistanceRunner: Represents a distance runner, using Athlete interface through composition.
    Jumper: Represents a jumper, using Athlete interface through composition.

Usage:
    Import the required classes and create instances of different types of athletes.
    Each athlete instance can display information about themselves using the `display()` method.


Author: Dac Linh Giao NGUYEN
Date: 2024

"""
from abc import abstractmethod


class AthleteInterface:
    """Interface defining the method for athlete specialty."""

    @abstractmethod
    def my_speciality(self):
        """Abstract method to be implemented by subclasses."""
        pass
